Feature/home top rated movies

Noor-Serry opened the pull request:
# Pull Request Template

## Description
implement home top rated movies section and its screen
---

## Changes Made
List the changes introduced in this pull request:

- add use case to get all top rated movies
- add use case to Compian the use cases 
- add the top rated movies screen and its view model and its effect and state , interaction listener and so on 
- write test for those use case with test coverage 100% (GetTopRatedMoviesUseCase , GetPopularMoviesUseCase and GetHomeScreenDataUseCase )

---

## Screenshots (if applicable)
Include screenshots or GIFs to showcase the changes, especially if they impact the UI.
https://www.veed.io/view/54f4f7e8-ba6f-4306-8bf8-dfaba9e30db0
<img width="939" height="34" alt="image" src="https://github.com/user-attachments/assets/6a31473a-ebde-46a2-9660-907c7c7f9436" />
<img width="943" height="41" alt="image" src="https://github.com/user-attachments/assets/cb25aeda-5558-4b06-ae59-ba186dfbdbfa" />
<img width="947" height="43" alt="image" src="https://github.com/user-attachments/assets/f67151df-8c6a-4abd-a130-07a38569639b" />

---

## Checklist
Please ensure the following tasks are completed:

- [x] My code follows the code style of this project
- [x] Changes have been tested manually and verified.
- [x] PR includes at most one single feature.

---

## Additional Comments
Add any additional information or context about the pull request here.

Commit:
feature: implement Home Screen and Top Rated Movies Screen

This commit introduces the Home Screen and the Top Rated Movies Screen, along with their respective functionalities and UI components.

**Key changes:**

**Domain Layer:**
- Added `GetHomeScreenDataUseCase` to fetch data for the home screen, including top-rated and popular movies.
- Added `GetPopularMoviesUseCase` and `GetTopRatedMoviesUseCase`.
- Updated `MovieRepository` interface to include methods for fetching popular and top-rated movies.

**Repository Layer:**
- Implemented methods in `MovieRepositoryImpl` to fetch popular and top-rated movies from the remote data source.

**Remote Datasource:**
- Added methods in `MovieRemoteDataSourceImpl` and `MovieRemoteSource` to fetch popular and top-rated movies from the API.

**ViewModel Layer:**
- Created `HomeViewModel` to manage the state and logic for the Home Screen.
- Created `TopRatedViewModel` for the Top Rated Movies Screen.
- Added corresponding UI state mappers (`HomeUiStateMapper`, `TopRatedUiStateMapper`), effects, and interaction listeners.

**UI Layer:**
- **Home Screen:**
    - Implemented `HomeScreen` with `HomeScreenContent`.
    - Displays sections for Popular movies and Top Rated movies.
    - Includes `HomeAppBar` for navigation and search.
    - Handles loading and error states.
    - Introduced `AnimatedSectionVisibility` for smooth transitions.
    - Added `TopRatingSection` for displaying top-rated movies.
- **Top Rated Movies Screen:**
    - Implemented `TopRatedScreen` with `TopRatedContent`.
    - Displays a grid of top-rated movies (`TopRatedMoviesGrid`).
    - Features a dynamic background (`TopRatedBackgroundComponent`) with blurred fire images.
    - Includes `DefaultAppBar` with dynamic background color based on scroll.
    - Handles loading and error states.
    - Added navigation for this screen.
- **Components:**
    - Created `TopRatedBackgroundComponent` for the visually appealing background.
    - Created `TopRatedMoviesGrid` to display movies in a grid.
- **Navigation:**
    - Added routes for Home and Top Rated screens in `NavGraph` and `Route`.
- **Theme:**
    - Added `topRatedGradientBackground` color to the theme.
- **Resources:**
    - Added `fire.png` drawable.

**DI (Koin):**
- Updated `useCaseModule` and `viewModelModule` to include dependencies for the new screens and use cases.

**Other:**
- Made `MovieImage` in `SearchByActorScreen` internal.
- Updated `ndkVersion` in `blurred/build.gradle.kts`.
- Removed background color from `LoadingContainer` for flexibility.

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into feature/home-top-rated-movies

# Conflicts:
#	app/src/main/java/com/amsterdam/aflami/di/useCaceModule.kt
#	app/src/main/java/com/amsterdam/aflami/di/viewModelModule.kt
#	domain/src/main/java/com/example/domain/repository/MovieRepository.kt
#	remoteDatasource/src/main/java/com/example/remotedatasource/datasource/MovieRemoteDataSourceImpl.kt
#	repository/src/main/java/com/example/repository/datasource/remote/MovieRemoteSource.kt
#	repository/src/main/java/com/example/repository/repository/MovieRepositoryImpl.kt
#	ui/src/main/java/com/example/ui/screens/home/HomeScreen.kt

Commit:
Refactor: remove redundant ndkVersion from build.gradle.kts

The `ndkVersion` property was removed from the `blurred/build.gradle.kts` file. This change relies on the NDK version specified in the `gradle.properties` file, which is the recommended approach for setting the NDK version.

Commit:
feature: integrate top Rated Movies into Home Screen

This commit integrates the Top Rated Movies section into the Home Screen.

Key changes include:
- Modifying `HomeViewModel` to use `GetHomeScreenDataUseCase` to fetch both popular and top-rated movies.
- Updating `HomeUiStateMapper` to map `HomeScreenData` (containing both popular and top-rated movies) to `HomeUiState`.
- Adding `TopRatedMovieItemUiState` to `HomeUiState` to represent top-rated movies.
- Adding new interactions and effects to `HomeInteractionListener` and `HomeEffect` for handling clicks on movies and navigating to the Top Rated Movies screen.
- Adjusting the `PopularSection` UI by changing `requiredHeight` to `height` for the image.
- Removing an unused import `androidx.compose.ui.graphics.Color` from `TopRatedBackgroundComponent.kt`.
- Removing an unused import `androidx.compose.foundation.background` from `LoadingContainer.kt`.
- Updating the AGP version in `libs.versions.toml`.

Noor-Serry committed:
Delete blurred/build.gradle.kts

Commit:
Merge remote-tracking branch 'origin/feature/home-top-rated-movies' into feature/home-top-rated-movies

# Conflicts:
#	blurred/build.gradle.kts

nabilBouzineDev commented on viewModel/src/main/java/com/example/viewmodel/topRated/TopRatedViewModel.kt:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    private fun onGetTopRatedMoviesSuccess(topRatedMovies: List<Movie>) {
+        updateState { topRatedUiStateMapper.toUiState(topRatedMovies) }
```
I think you should update the error to null and loading to false in the success scenario

Noor-Serry replied:
that already done here I update the whole screen state

Commit:
test: add tests for GetHomeScreenDataUseCase, GetTopRatedMoviesUseCase, and GetPopularMoviesUseCase

This commit introduces unit tests for the following use cases:

- `GetHomeScreenDataUseCase`:
    - Verifies successful data retrieval when both dependent use cases (`GetTopRatedMoviesUseCase` and `GetPopularMoviesUseCase`) return data.
    - Checks that exceptions are thrown correctly if either dependent use case fails.
    - Tests scenarios where one or both dependent use cases return empty lists.
- `GetTopRatedMoviesUseCase`:
    - Ensures it returns a list of top-rated movies when the repository provides them.
    - Confirms that it throws an exception if the repository call fails.
- `GetPopularMoviesUseCase`:
    - Ensures it returns a list of popular movies when the repository provides them.
    - Confirms that it throws an exception if the repository call fails.

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into feature/home-top-rated-movies

# Conflicts:
#	remoteDatasource/src/main/java/com/example/remotedatasource/datasource/MovieRemoteDataSourceImpl.kt

Commit:
refactor: move getTopRatedMovies to MovieServiceProvider

This commit refactors the `getTopRatedMovies` function by moving its implementation from `MovieRemoteDataSourceImpl` to `MovieServiceProviderImpl`.

The following changes were made:
- Added `getTopRatedMovies` to the `MovieApiService` interface and its corresponding endpoint constant.
- Added `getTopRatedMovies` to the `MovieServiceProvider` interface.
- Implemented `getTopRatedMovies` in `MovieServiceProviderImpl` to call the new API service method.
- Updated `MovieRemoteDataSourceImpl` to call `movieServiceProvider.getTopRatedMovies()`.
- Added `navigationBarsPadding` to the `TopRatedMoviesGrid` in `TopRatedScreen.kt` for better UI on devices with navigation bars.

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into feature/home-top-rated-movies

# Conflicts:
#	app/src/main/java/com/amsterdam/aflami/di/useCaceModule.kt
#	designSystem/src/main/java/com/example/designsystem/theme/colors/AflamiColorScheme.kt
#	designSystem/src/main/java/com/example/designsystem/theme/colors/DarkColors.kt
#	designSystem/src/main/java/com/example/designsystem/theme/colors/LigthColors.kt
#	domain/src/main/java/com/example/domain/repository/MovieRepository.kt
#	remoteDatasource/src/main/java/com/example/remotedatasource/datasource/MovieRemoteDataSourceImpl.kt
#	repository/src/main/java/com/example/repository/datasource/remote/MovieRemoteSource.kt
#	repository/src/main/java/com/example/repository/repository/MovieRepositoryImpl.kt
#	ui/src/main/java/com/example/ui/screens/home/HomeScreen.kt
#	viewModel/src/main/java/com/example/viewmodel/home/HomeEffect.kt
#	viewModel/src/main/java/com/example/viewmodel/home/HomeInteractionListener.kt
#	viewModel/src/main/java/com/example/viewmodel/home/HomeUiState.kt
#	viewModel/src/main/java/com/example/viewmodel/home/HomeViewModel.kt

Commit:
feature: implement blurred poster on home screen

This commit introduces a blurred movie poster background to the home screen.

Key changes include:

- **BlurredMoviePoster Composable:** A new composable `BlurredMoviePoster` has been created to display a blurred image with a gradient overlay.
- **HomeScreen Integration:**
    - The `HomeScreen` now uses `BlurredMoviePoster` to display the poster of the current popular movie.
    - The blur effect dynamically adjusts its vertical offset based on scroll interaction.
    - The AppBar background color transitions from transparent to the surface color based on scroll position.
- **ViewModel & UseCase Updates:**
    - `GetHomeScreenDataUseCase` now fetches upcoming movies in addition to popular and top-rated movies.
    - `HomeViewModel` is updated to use the new `GetHomeScreenDataUseCase` and handles the combined data.
    - Data fetching logic in `HomeViewModel` is streamlined to a single call for all home screen data.
- **UI State & Mapper Refactor:**
    - `TopRatedMovieItemUiState` has been replaced by a shared `MovieItemUiState`.
    - Mappers (`HomeUiStateMapper`, `TopRatedUiStateMapper`) are updated to use the shared `MovieItemUiState`.
- **Component Adjustments:**
    - `PopularMovieCard` now accepts `ratingAlpha`, `imageWidth`, and `imageHeight` to allow for dynamic visual changes.
    - Minor layout adjustments in `TopRatingSection` and `PopularSection`.
- **Navigation:**
    - Corrected navigation effect names in `HomeEffect` (e.g., `NavigateToMovieDetailsEffect` instead of `NavigateToMovieDetailsScreen`).
- **Removed Unused Code:**
    - Removed `getPopularMovies` from `MovieRepositoryImpl` as it's now handled by `GetHomeScreenDataUseCase`.

Commit:
refactor: Center LoadingContainer in HomeScreen and update tests

The LoadingContainer in HomeScreen is now centered using a `Box` with `contentAlignment = Alignment.Center`.

The `GetHomeScreenDataUseCaseTest` has been updated to include tests for `getUpcomingMoviesUseCase` and ensure that all three data sources (top-rated, popular, and upcoming movies) are correctly handled. Test names have been made more descriptive. A redundant comment was also removed from `MovieServiceProviderImpl.kt`.

Belal-Alkhatib commented on ui/src/main/java/com/example/ui/screens/home/sections/TopRatingSection.kt:
```diff
@@ -0,0 +1,55 @@
+import com.example.ui.screens.search.actorSearch.MovieImage
+import com.example.viewmodel.shared.uiStates.MovieItemUiState
+
+fun LazyListScope.TopRatingSection(
```
this fun name should start with lowercase just for kotlin ruls and consistency.

Belal-Alkhatib commented on ui/src/main/java/com/example/ui/screens/home/HomeScreen.kt:
```diff
@@ -63,44 +92,99 @@ private fun HomeScreenContent(
+                )
+            }
+        }
+        AnimatedSectionVisibility(visible = state.error != null) {
+            NoNetworkContainer(
+                onClickRetry = interactionListener::onClickRetryLoading
+            )
+        }
```
this part is already handled in homeErrorContent part see line 179, so you can remove it

Belal-Alkhatib commented on ui/src/main/java/com/example/ui/screens/home/HomeScreen.kt:
```diff
@@ -63,44 +92,99 @@ private fun HomeScreenContent(
+                )
+            }
+        }
+        AnimatedSectionVisibility(visible = state.error != null) {
```
this part is already handled in homeLoadingContent part see line 184, so you can remove it

Belal-Alkhatib commented on ui/src/main/java/com/example/ui/screens/home/HomeScreen.kt:
```diff
@@ -63,44 +92,99 @@ private fun HomeScreenContent(
+    val nestedScrollConnection = remember {
+        object : NestedScrollConnection {
+            override fun onPreScroll(available: Offset, source: NestedScrollSource): Offset {
+                val newOffset = blurOffsetY + available.y
+                blurOffsetY = newOffset.coerceIn(minimumValue = -1000f, maximumValue = -12f)
+                return Offset.Zero
+            }
+        }
+    }
+
+    Box(modifier = Modifier
+        .fillMaxSize()
+        .nestedScroll(nestedScrollConnection)) {
+        AnimatedSectionVisibility(visible = state.popularMovies.isNotEmpty()) {
+            BlurredMoviePoster(
+                posterUrl = state.popularMovies[pagerState.currentPage % state.popularMovies.size].posterUrl,
+                modifier = Modifier.offset { IntOffset(x = 0, y = blurOffsetY.roundToInt()) }
+            )
+        }
```
to make things consistent and code easy to read, please extract this to another section or supfucntion and call it here

Noor-Serry replied:
Here I just call component

Belal-Alkhatib replied:
most screens are. My point was just about keeping the UI code organized and easy to maintain. Extracting it into a small function helps with readability and consistency across the project

Belal-Alkhatib commented on ui/src/main/java/com/example/ui/screens/topRated/TopRatedScreen.kt:
```diff
@@ -0,0 +1,134 @@
+            it?.let {
+                when (it) {
+                    is TopRatedEffect.NavigateToMovieDetailsScreen -> {
+                        navController.navigate(Route.MovieDetails(it.movieId))
```
use safeNafigate here please

Belal-Alkhatib commented on ui/src/main/java/com/example/ui/screens/topRated/TopRatedScreen.kt:
```diff
@@ -0,0 +1,134 @@
+            AnimatedSectionVisibility(
+                visible = state.isLoading
+            ) {
+                Box(
+                    modifier = Modifier.fillMaxSize(),
+                    contentAlignment = Alignment.Center
```
unnecessary box usage here

Noor-Serry replied:
no it's necessary

Belal-Alkhatib commented on ui/src/main/java/com/example/ui/screens/topRated/TopRatedScreen.kt:
```diff
@@ -0,0 +1,134 @@
+            AnimatedSectionVisibility(
+                visible = state.error == TopRatedUiState.TopRatedError.NetworkError
+            ) {
+                Box(
+                    modifier = Modifier.fillMaxSize().zIndex(10f),
+                    contentAlignment = Alignment.Center
+                ) {
```
same here

Noor-Serry replied:
no it's necessary

Commit:
refactor: rename TopRatingSection to topRatingSection

The `TopRatingSection` composable function and its usage have been renamed to `topRatingSection` to follow Kotlin's lowerCamelCase convention for function names.

Additionally, the `HomeScreen` was updated to remove the `homeErrorContent` and `homeLoadingContent` as these are handled by the `MainScreen`. The navigation in `TopRatedScreen` was updated to use `safeNavigate`.

Commit:
refactor: simplify loading and error handling in HomeScreen

This commit removes the `homeErrorContent` and `homeLoadingContent` functions from `HomeErrorAndLoadingContent.kt` and integrates their logic directly into `HomeScreen.kt`.

The `HomeScreen` now directly uses `LoadingContainer` and `NoNetworkContainer` with `AnimatedSectionVisibility` to display loading and error states. The `zIndex` for the `LoadingContainer` has also been removed.

Commit:
refactor: simplify loading and error states in TopRatedScreen

This commit refactors the `TopRatedScreen` to simplify the UI for loading and network error states.

- The `Box` wrappers for `LoadingContainer` and `NoNetworkContainer` have been removed, and their modifiers are now directly applied to the respective containers.
- The `zIndex` for `NoNetworkContainer` has been removed as it's not necessary when only one of the error/loading states is visible at a time due to `AnimatedSectionVisibility`.

eslam-99 commented on designSystem/src/main/java/com/example/designsystem/components/LoadingContainer.kt:
```diff
@@ -15,7 +14,7 @@ import com.example.designsystem.theme.AppTheme
 fun LoadingContainer(modifier: Modifier = Modifier) {
     Column(
-        modifier = modifier.background(AppTheme.color.surface),
+        modifier = modifier,
```
Did you make sure this update in design system won't affect any component depend on it?

Noor-Serry replied:
yes the loading indictor should not have background

eslam-99 commented on repository/src/main/java/com/example/repository/repository/MovieRepositoryImpl.kt:
```diff
@@ -233,4 +229,10 @@ class MovieRepositoryImpl(
     }
-}
\ No newline at end of file
+    override suspend fun getPopularMovies(): List<Movie> =
```
Please use normal not compact functions for consistency, and also move them up in this file to gather all overrides together for better organization

Noor-Serry replied:
its normal to have compact here its one line

eslam-99 replied:
I know but there won't be consistency, I remember Bareq mentioned this to try to follow one approach

Noor-Serry replied:
id not remember that bareq said to do not make compact function even if it one line . I remember he said do not do compact function if its not compact ( have multiple lines )

eslam-99 replied:
Maybe, no problem

eslam-99 commented on ui/src/main/java/com/example/ui/screens/topRated/TopRatedScreen.kt:
```diff
@@ -0,0 +1,124 @@
+}
+
+@Composable
+fun TopRatedContent(
```
This can be private

eslam-99 commented on ui/src/main/res/drawable/fire.png:
why not adding this drawable for various sizes

eslam-99 commented on viewModel/src/main/java/com/example/viewmodel/home/HomeViewModel.kt:
```diff
@@ -1,56 +1,60 @@
         tryToExecute(
-            action = { getPopularMoviesUseCase() },
-            onSuccess = ::onGetPopularMovieSuccess,
+            action = { getHomeScreenDataUseCase() },
```
`action = ::getHomeScreenDataUseCase` for consistency

Noor-Serry replied:
that will not work

eslam-99 replied:
What about `getPopularMoviesUseCase::invoke`, if no need you can resolve the conversation

Noor-Serry replied:
not work

Commit:
refactor: minor code style and resource improvements

This commit includes the following changes:

- Changed the visibility of `TopRatedContent` composable function to `private` in `TopRatedScreen.kt`.
- Reordered function definitions in `MovieRepositoryImpl.kt` for better readability.
- Moved `fire.png` from `ui/src/main/res/drawable` to density-specific drawable folders (`drawable-hdpi`, `drawable-mdpi`, etc.).
